#### Q1. Pandas version

What's the version of Pandas that you installed?

In [1]:
import pandas as pd
pd.__version__

'2.0.3'

#### Getting the data
We'll use the Laptops Price dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv) to your local drive

Now read it with Pandas.

In [2]:
df = pd.read_csv('laptops.csv')
df


Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

### Q2. Records count

How many records are in the dataset?

- 12
- 1000
- 2160
- 12160

In [ ]:
### Get number of rows in Dataframe
len(df)

2160

Answer: 2160

### Q3. Laptop brands

How many laptop brands are presented in the dataset?

- 12
- 27
- 28
- 2160

In [ ]:
### No of unique brands
df['Brand'].nunique()

27

In [ ]:
### No of unique brands (dropna = true means exclude NaN in the count)
df['Brand'].nunique(dropna=True)

27

Answer: 27

### Q4. Missing values

How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3

In [ ]:
### Get series of columns and their null values
series = df.isnull().sum()
series

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

In [ ]:
### filter the series to have columns having null values and get the length of filtered series
len(series[series > 0])

3

Answer: 3

### Q5. Maximum final price

What's the maximum final price of Dell notebooks in the dataset?

- 869
- 3691
- 3849
- 3936

In [ ]:
### Filter on Brand == DELL. Access the Final Price column and apply max() on it
df[df.Brand.str.upper() == 'DELL']['Final Price'].max()

3936.0

Answer: 3936

### Q6. Median value of Screen

1. Find the median value of `Screen` column in the dataset.
2. Next, calculate the most frequent value of the same `Screen` column.
3. Use `fillna` method to fill the missing values in `Screen` column with the most frequent value from the previous step.
4. Now, calculate the median value of `Screen` once again.

Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- No

In [23]:
df.Screen.describe()

count    2156.000000
mean       15.168112
std         1.203329
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.000000
Name: Screen, dtype: float64

In [36]:
most_freq_screen = df.Screen.mode()
most_freq_screen

0    15.6
Name: Screen, dtype: float64

In [27]:
### Before replacing NA values with most frequent screen value
df.Screen.median()

15.6

In [ ]:
### Null values in screen column Before replacing with most freq screen value
df.Screen.isnull().sum()

4

In [ ]:
### Fill NaN values with most freq screen value (value for index 0) and re-assign to Screen column
df.Screen = df.Screen.fillna(most_freq_screen[0])

In [ ]:
### CHECK: Now, there should not be any NaN value in Screen. Following should return 0
df.Screen.isnull().sum()

0

In [ ]:
### Median After replacing NaN values in Screen.
df.Screen.median()
### No change

15.6

Answer: No change in Median

### Q7. Sum of weights

1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns `RAM`, `Storage`, `Screen`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the sum of all the elements of the result?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.43
- 45.29
- 45.58
- 91.30

In [ ]:
### Filter on brand = INNJOO, and select only 3 specific columns. Convert to Numpy Array. Call this X
X = df[df.Brand.str.upper() == 'INNJOO'][['RAM','Storage','Screen']].to_numpy()
X
# Matrix 6x3

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [46]:
Xt = X.transpose()
Xt
# Matrix 3x6

array([[  8. ,   8. ,   4. ,   6. ,   6. ,   6. ],
       [256. , 512. ,  64. ,  64. , 128. , 128. ],
       [ 15.6,  15.6,  14.1,  14.1,  14.1,  14.1]])

In [ ]:
Xtx = Xt.dot(X)